common/certificate/src/lib.rs

```diff
@@ -0,0 +1,355 @@
+use chrono::offset::Utc;
+use chrono::DateTime;
+use chrono::Duration;
+use rcgen::Certificate;
+use rcgen::CertificateParams;
+use rcgen::RcgenError;
+use sha1::{Digest, Sha1};
+use zeroize::Zeroizing;
+
+pub struct PemCertificate {
+    pem: x509_parser::pem::Pem,
+}
+
+impl PemCertificate {
+    pub fn from_pem_file(path: &str) -> Result<PemCertificate, CertificateError> {
+        let file = std::fs::File::open(path)?;
+        let (pem, _) = x509_parser::pem::Pem::read(std::io::BufReader::new(file))?;
+        Ok(PemCertificate { pem })
+    }
+
+    pub fn from_pem_string(content: &str) -> Result<PemCertificate, CertificateError> {
+        let file = std::io::Cursor::new(content.as_bytes());
+        let (pem, _) = x509_parser::pem::Pem::read(std::io::BufReader::new(file))?;
+        Ok(PemCertificate { pem })
+    }
+
+    pub fn subject(&self) -> Result<String, CertificateError> {
+        let x509 = PemCertificate::extract_certificate(&self.pem)?;
```

Comment:
After seeing this repeated "extraction" done in each and every getter methods like `subject`, `subject_common_name` etc, I'm wondering if this `X509Certificate` form of the certificate can be kept directly in the `PemCertificate` instead of the `Pem` form, provided the thumprint can also be generated from it.

Reply:
Good point! I tried but the solution is not obvious because the `X509Certificate<`a`> hides a reference to the `Pem` struct.

=> I will see with @mneumann how this can be improved.

Reply:
@didier-wenzek @albinsuresh You cannot easily have something like this:

```rust
pub struct PemCertificate {
    pem: x509_parser::pem::Pem,
    x509: &typeof(PemCertificate::extract_certificate) // substitute the type.
}
```

Where `x509` points to inside of `pem`.

Because this would be a self-referential struct. By default any struct is movable (unless it's pinned). If you would move the above struct, you would first allocate memory to hold a new `PemCertificate` and then do a `memcpy` of the contents (you cannot change this in Rust). But as the `x509` field points to a memory location within the same struct, you'd get a stale pointer, as you basically just copied the same address, which would obviously still point to the old memory location. You can `Pin` that structure but it's really NOT worth it!

## What you can do instead is:

```rust
/// This is the "owned" type.
pub struct PemCertificate {
    pem: x509_parser::pem::Pem,
}

/// This provides a readonly view / access to `subject_common_name` etc.
pub struct PemCertificateReader {
  pem: &x509_parser::pem::Pem,
  x509: &...
}

impl  PemCertificate {
  fn as_reader(&self) -> Result<PemCertificateReader, CertificateError> {
    let x509 = PemCertificate::extract_certificate(&self.pem)?;
    Ok(PemCertificateReader {
        pem: &self.pem,
        x509,
    })
  }
}

impl PemCertificateReader {
    pub fn subject_common_name(&self) {
       // ...
    }
}

```



And now the caller has to call  `pem.as_reader()?.subject_common_name()` instead of just `pem.subject_common_name()`.

Reply:
Note that `extract_certificate` is returning a `Result<x509_parser::certificate::X509Certificate<'a>, _>`. That `'a` *does exist* in realit,y but one does not have to write it out as Rust will infer it automatically from `&pem`.

You can explicitly specify the lifetime:

```rust
fn extract_certificate<'a>(pem: &'a x509_parser::pem::Pem) -> 
   Result<x509_parser::certificate::X509Certificate<'a>, CertificateError> {
  // ...
}
```

Which reads: `forall` lifetimes `'a` define function `extract_certificate` that takes an argument `&pem` that has to live at least as long as `'a` and returns a reference that MAY NOT outlive `'a`. 

This is what Rust infers for you.

Reply:
Thanks @mneumann.

However, I will not change the current code. Because, wrapping the `x509_parser::certificate::X509Certificate<'a>` into a `PemCertificateReader<'a>` just push the issue one step further. One still has to build such a reader in each function. And that has little value here since wrapping a single value.